File: desktop/native/SurveyController.App/Services/RpcServices.cpp
#include "pch.h"
#include "RpcServices.h"
#include "BackendClient.h"

namespace winrt::SurveyController::App::Services
{
    namespace
    {
        hstring StringRequest(wchar_t const* key, hstring const& value)
        {
            Windows::Data::Json::JsonObject request;
            request.SetNamedValue(key, Windows::Data::Json::JsonValue::CreateStringValue(value));
            return request.Stringify();
        }
    }

    Windows::Foundation::IAsyncOperation<hstring> SettingsService::LoadAsync() const
    { co_return co_await BackendClient::Current().CallAsync(L"GetAppSettings"); }
    Windows::Foundation::IAsyncOperation<hstring> SettingsService::SaveAsync(hstring request) const
    { co_return co_await BackendClient::Current().CallAsync(L"SaveAppSettings", request); }
    Windows::Foundation::IAsyncOperation<hstring> SettingsService::ResetAsync() const
    { co_return co_await BackendClient::Current().CallAsync(L"ResetAppSettings"); }

    Windows::Foundation::IAsyncOperation<hstring> ConfigService::LoadAsync(hstring path) const
    { co_return co_await BackendClient::Current().CallAsync(L"LoadConfig", path.empty() ? L"{}" : StringRequest(L"path", path)); }
    Windows::Foundation::IAsyncOperation<hstring> ConfigService::SaveAsync(hstring request) const
    { co_return co_await BackendClient::Current().CallAsync(L"SaveConfig", request); }
    Windows::Foundation::IAsyncOperation<hstring> ConfigService::CreateSurveyAsync(hstring url) const
    { co_return co_await BackendClient::Current().CallAsync(L"CreateSurveyDocument", StringRequest(L"url", url)); }
    Windows::Foundation::IAsyncOperation<hstring> ConfigService::DecodeQrCodeAsync(hstring path) const
    { co_return co_await BackendClient::Current().CallAsync(L"DecodeQRCode", StringRequest(L"path", path)); }

    Windows::Foundation::IAsyncOperation<hstring> TaskService::CheckAsync(hstring request) const
    { co_return co_await BackendClient::Current().CallAsync(L"CheckTask", request); }
    Windows::Foundation::IAsyncOperation<hstring> TaskService::StartAsync(hstring request) const
    { co_return co_await BackendClient::Current().CallAsync(L"StartRun", request); }
    Windows::Foundation::IAsyncOperation<hstring> TaskService::StateAsync(hstring runId, std::uint64_t afterSequence) const
    {
        Windows::Data::Json::JsonObject request;
        request.SetNamedValue(L"runId", Windows::Data::Json::JsonValue::CreateStringValue(runId));
        request.SetNamedValue(L"afterSequence", Windows::Data::Json::JsonValue::CreateNumberValue(static_cast<double>(afterSequence)));
        co_return co_await BackendClient::Current().CallAsync(L"GetRunTaskState", request.Stringify(), std::chrono::seconds{ 5 });
    }
    Windows::Foundation::IAsyncOperation<hstring> TaskService::PauseAsync(hstring reason) const
    { co_return co_await BackendClient::Current().CallAsync(L"PauseRun", StringRequest(L"value", reason)); }
    Windows::Foundation::IAsyncOperation<hstring> TaskService::ResumeAsync() const
    { co_return co_await BackendClient::Current().CallAsync(L"ResumeRun"); }
    Windows::Foundation::IAsyncOperation<hstring> TaskService::StopAsync() const
    { co_return co_await BackendClient::Current().CallAsync(L"CancelRun"); }
    Windows::Foundation::IAsyncOperation<hstring> TaskService::ExportAsync(hstring path, Windows::Data::Json::JsonArray lines) const
    {
        Windows::Data::Json::JsonObject request;
        request.SetNamedValue(L"path", Windows::Data::Json::JsonValue::CreateStringValue(path));
        request.SetNamedValue(L"lines", lines);
        co_return co_await BackendClient::Current().CallAsync(L"ExportLogLines", request.Stringify());
    }
    Windows::Foundation::IAsyncOperation<hstring> TaskService::TestAiAsync(Windows::Data::Json::JsonObject profile) const
    {
        Windows::Data::Json::JsonObject request;
        request.SetNamedValue(L"aiProfile", profile);
        co_return co_await BackendClient::Current().CallAsync(L"TestAIConnection", request.Stringify());
    }

    Windows::Foundation::IAsyncOperation<hstring> ProxyService::AreasAsync(hstring source) const
    { co_return co_await BackendClient::Current().CallAsync(L"GetProxyAreaOptions", StringRequest(L"value", source)); }
    Windows::Foundation::IAsyncOperation<hstring> ProxyService::TestFixedAsync(hstring address) const
    { co_return co_await BackendClient::Current().CallAsync(L"TestFixedProxy", StringRequest(L"address", address)); }
    Windows::Foundation::IAsyncOperation<hstring> ProxyService::TestCustomAsync(hstring url) const
    { co_return co_await BackendClient::Current().CallAsync(L"TestCustomProxyAPI", StringRequest(L"url", url)); }
    Windows::Foundation::IAsyncOperation<hstring> ProxyService::SyncAsync(hstring source) const
    { co_return co_await BackendClient::Current().CallAsync(L"SyncProxyStatus", StringRequest(L"value", source)); }

    Windows::Foundation::IAsyncOperation<hstring> CommunityService::CheckUpdateAsync(hstring currentVersion) const
    { co_return co_await BackendClient::Current().CallAsync(L"CheckUpdate", StringRequest(L"currentVersion", currentVersion)); }
    Windows::Foundation::IAsyncOperation<hstring> CommunityService::IpUsageAsync() const
    { co_return co_await BackendClient::Current().CallAsync(L"GetIPUsageSummary"); }
}

File: desktop/native-wpf/src/SurveyController.App/Services/RpcServices.cs
using SurveyController.Core.Rpc;

namespace SurveyController.App.Services;

/// <summary>
/// RPC 门面：页面只调用这里，不拼方法名或 JSON 信封。
/// 参数/返回值沿用 C++ 侧约定——字符串形式的 JSON。
/// </summary>
public static class SettingsService
{
    public static Task<string> LoadAsync() =>
        BackendClient.Current.CallAsync(RpcMethods.GetAppSettings, "null");

    public static Task<string> SaveAsync(string requestJson) =>
        BackendClient.Current.CallAsync(RpcMethods.SaveAppSettings, requestJson);

    public static Task<string> ResetAsync() =>
        BackendClient.Current.CallAsync(RpcMethods.ResetAppSettings, "null");
}

public static class ConfigService
{
    public static Task<string> LoadAsync(string? path = null)
    {
        var request = path is null ? new object() : new { path };
        return BackendClient.Current.CallAsync(RpcMethods.LoadConfig, Serialize(request));
    }

    public static Task<string> SaveAsync(string requestJson) =>
        BackendClient.Current.CallAsync(RpcMethods.SaveConfig, requestJson);

    public static Task<string> CreateSurveyAsync(string url) =>
        BackendClient.Current.CallAsync(RpcMethods.CreateSurveyDocument, Serialize(new { url }));

    public static Task<string> DecodeQRCodeAsync(string path) =>
        BackendClient.Current.CallAsync(RpcMethods.DecodeQRCode, Serialize(new { path }));

    public static Task<string> DecodeQRCodeSurveyAsync(string path) =>
        BackendClient.Current.CallAsync(RpcMethods.DecodeQRCodeSurvey, Serialize(new { path }));

    private static string Serialize(object value) =>
        System.Text.Json.JsonSerializer.Serialize(value);
}

public static class TaskService
{
    /// <summary>GetRunTaskState 轮询使用更短超时，与 C++ 侧 5s 一致。</summary>
    public static readonly TimeSpan RunStateTimeout = TimeSpan.FromSeconds(5);

    public static Task<string> CheckTaskAsync(string requestJson) =>
        BackendClient.Current.CallAsync(RpcMethods.CheckTask, requestJson);

    public static Task<string> StartRunAsync(string requestJson) =>
        BackendClient.Current.CallAsync(RpcMethods.StartRun, requestJson);

    public static Task<string> CheckAndStartAsync(string requestJson) =>
        BackendClient.Current.CallAsync(RpcMethods.CheckAndStart, requestJson);

    public static Task<string> GetRunTaskStateAsync(string runId, ulong afterSequence)
    {
        var payload = new System.Text.Json.Nodes.JsonObject
        {
            ["runId"] = runId ?? string.Empty,
            ["afterSequence"] = afterSequence,
        }.ToJsonString();
        return BackendClient.Current.CallAsync(RpcMethods.GetRunTaskState, payload, RunStateTimeout);
    }

    public static Task<string> PauseRunAsync(string reason) =>
        BackendClient.Current.CallAsync(RpcMethods.PauseRun, SerializeValue(new { value = reason }));

    public static Task<string> ResumeRunAsync() =>
        BackendClient.Current.CallAsync(RpcMethods.ResumeRun, "null");

    public static Task<string> CancelRunAsync() =>
        BackendClient.Current.CallAsync(RpcMethods.CancelRun, "null");

    public static Task<string> ExportLogLinesAsync(string path, IReadOnlyList<string> lines) =>
        BackendClient.Current.CallAsync(RpcMethods.ExportLogLines,
            SerializeValue(new { path, lines }));

    public static Task<string> TestAiConnectionAsync(string aiProfileJson) =>
        BackendClient.Current.CallAsync(RpcMethods.TestAIConnection,
            "{\"aiProfile\":" + aiProfileJson + "}");

    private static string SerializeValue<T>(T value) =>
        System.Text.Json.JsonSerializer.Serialize(value);
}

public static class ProxyService
{
    public static Task<string> GetProxyAreaOptionsAsync(string source) =>
        BackendClient.Current.CallAsync(RpcMethods.GetProxyAreaOptions,
            System.Text.Json.JsonSerializer.Serialize(new { value = source }));

    public static Task<string> TestFixedProxyAsync(string address) =>
        BackendClient.Current.CallAsync(RpcMethods.TestFixedProxy,
            System.Text.Json.JsonSerializer.Serialize(new { address }));

    public static Task<string> TestCustomProxyApiAsync(string url) =>
        BackendClient.Current.CallAsync(RpcMethods.TestCustomProxyAPI,
            System.Text.Json.JsonSerializer.Serialize(new { url }));

    public static Task<string> SyncStatusAsync(string source) =>
        BackendClient.Current.CallAsync(RpcMethods.SyncProxyStatus,
            System.Text.Json.JsonSerializer.Serialize(new { value = source }));
}

public static class CommunityService
{
    public static Task<string> CheckUpdateAsync(string currentVersion) =>
        BackendClient.Current.CallAsync(RpcMethods.CheckUpdate,
            System.Text.Json.JsonSerializer.Serialize(new { currentVersion }));

    public static Task<string> IpUsageAsync() =>
        BackendClient.Current.CallAsync(RpcMethods.GetIPUsageSummary, "null");
}
